overlord/state/state.go

```diff
@@ -362,12 +363,33 @@ func (s *State) tasksIn(tids []string) []*Task {
 
 // Prune removes changes that became ready for more than pruneWait
 // and aborts tasks spawned for more than abortWait.
-// It also removes tasks unlinked to changes after pruneWait.
-func (s *State) Prune(pruneWait, abortWait time.Duration) {
+// It also removes tasks unlinked to changes after pruneWait. When
+// there are more changes than the limit set via "maxReadyChanges"
+// those changes in ready state will also removed even if they are below
+// the pruneWait duration.
+func (s *State) Prune(pruneWait, abortWait time.Duration, maxReadyChanges int) {
 	now := time.Now()
 	pruneLimit := now.Add(-pruneWait)
 	abortLimit := now.Add(-abortWait)
-	for _, chg := range s.Changes() {
+
+	// sort from oldest to newest
+	changes := s.Changes()
+	sort.Sort(byReadyTime(changes))
+
+	// used just for couting
+	readyChanges := map[string]bool{}
```

Comment:
wondering, isn't a counter enough? also given the ordering can't we just count backward in the main loop (maybe a bit obscure)